posthog/models/filters/base_filter.py

```diff
@@ -43,7 +50,7 @@ def toJSON(self):
 
     def with_data(self, overrides: Dict[str, Any]):
         "Allow making copy of filter whilst preserving the class"
-        return type(self)(data={**self._data, **overrides}, **self.kwargs)
+        return type(self)(data={**self._data, **overrides}, hogql_context=self.hogql_context, **self.kwargs)
```

Comment:
I wrote above "Note that calling filter.with_data (cloning a filter) passes the context by reference, meaning both the old and new filter will gather variables into the same dict."

That's on purpose, as you might compose your query from multiple subqueries, and you'd want to share the variable collector between those and the host query. This happens with e.g. retention.

Reply:
🤯 that feels like a difficult to diagnose bug waiting to happen... coupling on instance is the most evil form of coupling 😰 

At the least the comment in `with_data` needs updating so that intellisense/scanning the code makes it clearer that the instance is shared. It's not making a copy anymore if we have `Filter instance A` <-> `hogql instance` <-> `Filter instance B`

Reply:
Open to suggestions on how to improve this. I suggested `shallow_clone` below, and [here's a PR](https://github.com/PostHog/posthog/pull/13918) that changes `with_data` to `shallow_clone`. I made a new PR since adding another 41 changed files to this one seems... audacious. 🤣